### PR TITLE
fix(evm): use forked sharedbackend correctly

### LINKED ISSUE #1026: Spawn backendhandler only once in fork mode
### Component

Forge

### Have you ensured that all of these are up to date?

- [ ] Foundry
- [ ] Foundryup

### What version of Foundry are you on?

master

### What command(s) is the bug in?

forge test

### Operating System

_No response_

### Describe the bug

currently multiple handler can be spawned

https://github.com/gakonst/foundry/blob/master/forge/src/multi_runner.rs#L219

instead a new `BackendHandler` should be spawned once and the tests should be run with with a clone of the `SharedBackend`

### PR DESCRIPTION
**Pull request by mattsse**
## Motivation
Closes #1026

## Solution
spawn a `BackendHandler` only once and clone the `SharedBackend`s instead while running tests in parallel

### CODE REVIEW

File: cli/src/cmd/forge/test.rs [hunk 1]
```diff
@@ -199,6 +199,7 @@ impl Cmd for TestArgs {
             self.filter.test_pattern = self.debug;
             match runner.count_filtered_tests(&self.filter) {
                 1 => {
+
```
Comment (onbjerg) [inline cli/src/cmd/forge/test.rs:202]:
```suggestion

```

**Review by onbjerg — APPROVED** [review]
smallest nit ever otherwise LGTM

File: cli/src/cmd/forge/run.rs [hunk 2]
```diff
@@ -96,11 +96,15 @@ impl Cmd for RunArgs {
...
+        let env = evm_opts.evm_env();
+        // the db backend that serves all the data
+        let db = Backend::new(utils::get_fork(&evm_opts, &config.rpc_storage_caching), &env);
+
         let mut builder = ExecutorBuilder::new()
             .with_cheatcodes(evm_opts.ffi)
-            .with_config(evm_opts.evm_env())
-            .with_spec(crate::utils::evm_spec(&config.evm_version))
-            .with_fork(utils::get_fork(&evm_opts, &config.rpc_storage_caching));
+            .with_config(env)
```
Comment (onbjerg) [inline cli/src/cmd/forge/run.rs:105]: Oh also, unrelated, but I forgot to call `with_gas_limit` here which may be relevant if `forge run` forks: can you call `with_gas_limit(evm_opts.gas_limit())` as well?

**Comment by mattsse** [comment]
can't reproduce the clippy failure locally, seems to be originating in another dependency

**Comment by onbjerg** [comment]
wow that's weird

**Review by gakonst — APPROVED** [review]
Spells Mainnet bench went to 9:50 from 12:13, gg

**Review by univerz — COMMENTED** [review]
i'm late to the party, but discovered some things when porting from sputnik & i'm not sure how to fix it myself properly.

File: evm/src/executor/fork/backend.rs [hunk 3]
```diff
@@ -235,100 +238,121 @@ where
...
-            pin.pending_requests.push(request);
         }
 
-        // the handler is finished if the request channel was closed and all requests are processed
-        if pin.incoming.is_done() && pin.pending_requests.is_empty() {
-            if let Poll::Ready(Some(ack)) = Pin::new(&mut pin.shutdown).poll_next(cx) {
-                // effectively flushing the cached storage if any
+        if let Poll::Ready(Some(ack)) = Pin::new(&mut pin.shutdown).poll_next(cx) {
+            trace!(target: "backendhandler", "received shutdown signal");
+            if pin.incoming.is_done() {
```
[the remark this answers is not in the document]
Comment (mattsse) [reply evm/src/executor/fork/backend.rs:337]: there was a problem because we never joined the backgroundhandler to the main thread but added cache flushing to it. which should be triggered if all senders are dropped but we realised we need to and wait on drop until the cache was flushed.

File: evm/src/executor/fork/backend.rs [hunk 4]
```diff
@@ -370,11 +394,13 @@ impl Drop for SharedBackend {
         // disconnect the command channel
         self.backend.disconnect();
         if self.backend.is_closed() {
```
[the remark this answers is not in the document]
Comment (mattsse) [reply evm/src/executor/fork/backend.rs:396]: you're right, this is redundant but hard to parse from the sender's API docs.
so this if check can be removed.

File: evm/src/executor/fork/backend.rs [hunk 5]
```diff
@@ -235,100 +238,121 @@ where
...
 
-            if pin.shutdown.is_done() {
-                trace!(target: "backendhandler", "finished");
-                return Poll::Ready(())
-            }
+        // the handler is finished if the request channel was closed and all requests are processed
+        if pin.incoming.is_done() &&
+            pin.pending_requests.is_empty() &&
+            pin.shutdown.is_done() &&
+            pin.shutdown.is_done()
```
[the remark this answers is not in the document]
Comment (mattsse) [reply evm/src/executor/fork/backend.rs:350]: good catch, 
need to check incoming, pending and shutdon, the second shutdown is done call is of course redundant

File: evm/src/executor/fork/backend.rs [hunk 6]
```diff
@@ -235,100 +238,121 @@ where
...
+            trace!(target: "backendhandler", "received shutdown signal");
+            if pin.incoming.is_done() {
+                // if request channel was closed, flush the cached storage to disk, if caching is
+                // enabled
                 pin.db.cache().flush();
-                // signaling back
-                let _ = ack.send(());
             }
+            // signaling back
+            let _ = ack.send(());
```
[the remark this answers is not in the document]
Comment (mattsse) [reply evm/src/executor/fork/backend.rs:343]: > the problem i ran into is: if SharedBackend is cloned, then BackendHandler is not dopped (=> leaking memory)

this is indented behaviour because we want to be able to use multiple `SharedBackend`s connected to the same handler.

here is the chain of events that will cause the `BackendHandler` to exit.

1. last SharedBackend is dropped
2. `incoming` channel is closed, 
3. SharedBackend sends `shutdown`
4. `BackendHandler` receives shutdown 
5. `incoming.is_done == true` so cache is flushed
6. send ack back to  SharedBackend
7. SharedBackend dropped closing the shutdown channel
8. BackendHandler enters exit condition and returns Ready

But we could return ready if it was the last SharedBackend that sent the shutdown msg

[the remark this answers is not in the document]
Comment (univerz) [reply evm/src/executor/fork/backend.rs:343]: > 8. BackendHandler enters exit condition and returns Ready

after `BackendHandler::poll` returns `Poll::Pending`, there is nothing scheduled to wake it (call poll again) to finally exit & drop.

[the remark this answers is not in the document]
Comment (mattsse) [reply evm/src/executor/fork/backend.rs:343]: that's interesting because I'd assume dropping the last shutdown sender(closing the channel) should awake the receiver's weaker.
just tested this and it's indeed inconsistent which is weird...
so the proper fix would be to exit after flushing the cache then

[the remark this answers is not in the document]
Comment (univerz) [reply evm/src/executor/fork/backend.rs:343]: > that's interesting because I'd assume dropping the last shutdown sender(closing the channel) should awake the receiver's waker.

yes, i didn't know how does it work either - it registers for 1 poll "iteration" & in this case there was no need to register again (value was available).
```rust
/// Attempt to pull out the next value of this stream, registering the
/// current task for wakeup if the value is not yet available
fn poll_next(...)
```

[the remark this answers is not in the document]
Comment (univerz) [reply evm/src/executor/fork/backend.rs:343]: btw, while we are at discussing backend - would you support this change? https://github.com/gakonst/foundry/pull/1003#discussion_r838435937

[the remark this answers is not in the document]
Comment (mattsse) [reply evm/src/executor/fork/backend.rs:343]: supportive1
